streams/src/main/java/org/apache/kafka/streams/processor/internals/PunctuationSchedule.java

```diff
@@ -16,30 +16,73 @@
  */
 package org.apache.kafka.streams.processor.internals;
 
+import org.apache.kafka.streams.processor.Cancellable;
+import org.apache.kafka.streams.processor.Punctuator;
+
 public class PunctuationSchedule extends Stamped<ProcessorNode> {
 
     final long interval;
+    final Punctuator punctuator;
+    boolean isCancelled = false;
+    // this Cancellable will be re-pointed at the successor schedule in next()
+    private final RepointableCancellable cancellable;
+
+    public PunctuationSchedule(ProcessorNode node, long interval, Punctuator punctuator) {
+        this(node, 0L, interval, punctuator);
+    }
 
-    public PunctuationSchedule(ProcessorNode node, long interval) {
-        this(node, 0L, interval);
+    public PunctuationSchedule(ProcessorNode node, long time, long interval, Punctuator punctuator) {
+        this(node, time, interval, punctuator, new RepointableCancellable());
+        cancellable.setSchedule(this);
     }
 
-    public PunctuationSchedule(ProcessorNode node, long time, long interval) {
+    private PunctuationSchedule(ProcessorNode node, long time, long interval, Punctuator punctuator, RepointableCancellable cancellable) {
```

Comment:
this one is also called from next() and the above one is called from StreamTask#schedule